### ISSUE TITLE
BoundedByteBuffer copies array every time when full.

### ISSUE THREAD

**Issue by SharplEr** [yegor256/cactoos-http#60]
In current implementation when `BoundedByteBuffer` is full then it going to copy full array on every `offer` operation. So complexity of `BoundedByteBuffer.offer()` is `O(n)`

But we can avoid any copy. Just look how `ArrayDeque` works inside:
```Java
    public void addFirst(E e) {
        if (e == null)
            throw new NullPointerException();
        elements[head = (head - 1) & (elements.length - 1)] = e;
        if (head == tail)
            doubleCapacity();
    }
```
Looks a bit complex but idea is very nice. They use one trick: `elements.length` is always a power of two. So `(head -1) & (elements.length - 1)` just drops higher bits of `head -1`. So when `head-1` become less than 0, then a `head` start to count from the `elements.length-1`. And so on in a circle.

In case if `elements.length` is not a power of two then we should use `x % elements.length` for making such circle.

Example in our case:
`elements = {0, 0, 0};`
`this.idx = 0;`
`this.full = false;`
1) Add 1:
`elements == {1, 0, 0}`
`this.idx = ((this.idx + 1) % 3) = 1;`
`this.full == false;`
2) Add 2:
`elements == {1, 2, 0}`
`this.idx = ((this.idx + 1) % 3) = 2;`
this.full == false;
3) Add 3:
elements == {1, 2, 3}
`this.idx = ((this.idx + 1) % 3) = 0;`
`this.full = true;`
4) Add 4:
`elements == {4, 2, 3}`
`this.idx = ((this.idx + 1) % 3) = 1;`
`this.full == true;`

**And then `offer` become `O(1)`.**

Then if we want to read there are two options:
1) If `this.full == false`: just read from `0` to `this.idx`,
2) If `this.full == true`: read from `this.idx+1` to `elements.size` then read from `0` to `this.idx`.

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) please, pay attention to this issue

**Comment by victornoel** [comment]
@llorllale for the record that's how it was meant to be implemented from the beginning: that's the whole idea behind using a bounded byte buffer.
@SharplEr in our case we don't even need to change capacity since it is of fixed size, so it is even simpler :)

**Comment by 0crat** [comment]
Job [#60](https://github.com/yegor256/cactoos-http/issues/60) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://www.zerocracy.com/policy.html#29): +15 point(s) just awarded to @sharpler[/z](https://www.0crat.com/u/sharpler)

**Comment by 0crat** [comment]
The job [#60](https://github.com/yegor256/cactoos-http/issues/60) assigned to @golyalpha[/z](https://www.0crat.com/u/golyalpha), here is [why](https://www.0crat.com/footprint/C63314D6Z/6f8128a3-3992-490c-8e21-a106e5d64017); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be no monetary reward for this job

**Comment by golyalpha** [comment]
@0crat waiting on #75

**Comment by 0crat** [comment]
> @0crat waiting on #75  ([here](https://github.com/yegor256/cactoos-http/issues/60#issuecomment-474578668))

@golyalpha The impediment for [#60](https://github.com/yegor256/cactoos-http/issues/60) was registered successfully by @golyalpha[/z](https://www.0crat.com/u/golyalpha)

**Comment by 0crat** [comment]
The job [#60](https://github.com/yegor256/cactoos-http/issues/60) assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/C63314D6Z/5925fad3-f663-4069-ae7a-c414fdf968dc); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be no monetary reward for this job

**Comment by llorllale** [comment]
@victornoel hold on

**Comment by victornoel** [comment]
@llorllale not sure what happened here, it is the second time I get assigned issues without the previous assignee refusing it...

**Comment by llorllale** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos-http/issues/60#issuecomment-475968305))

@llorllale This is what I know about this job in [C63314D6Z](https://www.0crat.com/p/C63314D6Z), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#60](https://github.com/yegor256/cactoos-http/issues/60) is [in scope](http://www.zerocracy.com/policy.html#14) for 9mon
  * The role is `DEV`
  * The job is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel) for [1days](http://www.zerocracy.com/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://www.zerocracy.com/policy.html#2) job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 30 minutes/points
  * These users are banned and won't be assigned:
    * @sharpler[/z](https://www.0crat.com/u/sharpler): This user reported the ticket
    * @golyalpha[/z](https://www.0crat.com/u/golyalpha): User was resigned from the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos-http%2360%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by llorllale** [comment]
@0crat assign @golyalpha

**Comment by llorllale** [comment]
@victornoel in this particular case, @golyalpha left the project and then applied again for some reason

**Comment by 0crat** [comment]
> @0crat assign @golyalpha ([here](https://github.com/yegor256/cactoos-http/issues/60#issuecomment-475968677))

@llorllale The job [#60](https://github.com/yegor256/cactoos-http/issues/60) assigned to @golyalpha[/z](https://www.0crat.com/u/golyalpha), here is [why](https://www.0crat.com/footprint/C63314D6Z/4c958fe1-ffa2-439d-b689-dbdaf39aeef4); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be no monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by golyalpha** [comment]
Yeah, sorry about that.

**Comment by 0pdd** [comment]
@SharplEr the puzzle [#77](https://github.com/yegor256/cactoos-http/issues/77) is still not solved.

**Comment by golyalpha** [comment]
@llorllale issue ready to close?

**Comment by llorllale** [comment]
@SharplEr can we close this? PR #75 was merged

**Comment by golyalpha** [comment]
@0crat waiting for issue author to respond

**Comment by 0crat** [comment]
> @0crat waiting for issue author to respond ([here](https://github.com/yegor256/cactoos-http/issues/60#issuecomment-478324758))

@golyalpha The impediment for [#60](https://github.com/yegor256/cactoos-http/issues/60) was registered successfully by @golyalpha[/z](https://www.0crat.com/u/golyalpha)

**Comment by golyalpha** [comment]
@llorllale Seems the issue author has been gone for a while. Can we deal with this some other way?

**Comment by llorllale** [comment]
@SharplEr closing issue due to inactivity. PR #75 was merged. Anything else, raise another issue.

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @golyalpha[/z](https://www.0crat.com/u/golyalpha), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Job [#60](https://github.com/yegor256/cactoos-http/issues/60) is not in the agenda of @golyalpha[/z](https://www.0crat.com/u/golyalpha), can't inspect

**Comment by 0crat** [comment]
The job [#60](https://github.com/yegor256/cactoos-http/issues/60) is now out of scope

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +30 point(s) just awarded to @golyalpha[/z](https://www.0crat.com/u/golyalpha)

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @sereshqua[/z](https://www.0crat.com/u/sereshqua)

**Comment by 0pdd** [comment]
@SharplEr the only puzzle [#77](https://github.com/yegor256/cactoos-http/issues/77) is solved here.
